### PR TITLE
Search and filter trial table

### PR DESCRIPTION
**Pull request by sylvr3**
Still a work in progress. Not ready to merge yet.

### CODE REVIEW

**Comment by sylvr3** [comment]
Whenever I use the dataTable function and the page loads it gives me the following error:

DataTables warning: table id=trials - Requested unknown parameter '1' for row 0. For more information about this error, please see http://datatables.net/tn/4

I am also getting this error for row 3.

I have tried creating a field called defaultContent and setting that to an empty string, as shown here:
http://datatables.net/reference/option/columns.defaultContent

I have also tried to use mRender to display the table in the event of a null value (see last example on this page):
http://stackoverflow.com/questions/27502267/how-to-handle-database-null-values-in-jquery-datatable

Any suggestions? @ChristianMurphy

**Comment by ChristianMurphy** [comment]
@sylvr3 Try printing the data being send in the console to be sure the property names sent are the same as DataTables is looking for.

**Comment by sylvr3** [comment]
If I revert back to commit 216d295 (that actually has property names), and console.log(data) for instance, nothing gets printed. Time to find another approach....

**Comment by ChristianMurphy** [comment]
Issue was that the current rendering uses two rows to render a trial which have different number of columns.

Combining the two rows into one resolved the error.
This is the current appearance.

datatable

**Comment by sylvr3** [comment]
Thanks Christian. :+1: 
I am looking into the filtering issue now.

**Comment by ChristianMurphy** [comment]
@sylvr3 the `data-search` and `data-order` attributes in the `<td>` element allow control over how Datatables will filter and search over the column.

**Comment by sylvr3** [comment]
@ChristianMurphy is the bootstrap dropdown menu disabled on your end? Should I use a custom search box to get it out of the way?

**Comment by ChristianMurphy** [comment]
> is the bootstrap dropdown menu disabled on your end?

Which dropdown menu are you referring to?

> Should I use a custom search box to get it out of the way?

The search box seems to work fine, probably not.

**Comment by sylvr3** [comment]
I'm referring to the Status dropdown next to Add New.

**Comment by ChristianMurphy** [comment]
It still works for me.
Since the drop down takes focus over the search box there shouldn't be any usability issues here.

dropdown

**Comment by sylvr3** [comment]
Hmmm...I pulled from this branch and master. I will try clearing my cache and see if that helps.

**Comment by sylvr3** [comment]
I have everything up to date and I am still getting that DataTables warning and blocked drop down...

**Comment by ChristianMurphy** [comment]
Okay first thing, check `git rev-parse HEAD`
It should be on commit `482b8eca3fa966bffc2d8542434ee20afeee60cc`

Second open the browser developer tools.
And look at the console and let me know what (if any) errors it is showing.

**Comment by sylvr3** [comment]
git rev-parse HEAD is giving me: e80a220f8d6473a1b4e6fda0e25913a5a15fc3ee

I am not getting any errors. I did a npm install -g because I was missing some new dependencies, but no errors in the console. I am still getting the DataTables warning alert box though.

**Comment by ChristianMurphy** [comment]
It still has the error because your local copy of the branch is outdated.
Pull the latest branch and test again.

**Comment by ChristianMurphy** [comment]
For the missing NPM package warning.
Delete the `node_modules` folder and run `npm install` again.

**Comment by sylvr3** [comment]
npm install -g did the trick.

**Comment by sylvr3** [comment]
Is the dropdown button click working on your end? I want to make sure the sample alert message appears successfully before I proceed further.

**Comment by ChristianMurphy** [comment]
Yep I see the alert when clicking on an option.

**Comment by sylvr3** [comment]
All done! :smile: 

screenshot from 2015-11-20 17 41 49
screenshot from 2015-11-20 17 38 48

screenshot from 2015-11-20 17 40 42

**Comment by sylvr3** [comment]
@ChristianMurphy Let me know if this looks good.

File: view/template/dashboard.hbs [hunk 1]
```diff
@@ -225,7 +229,63 @@
         data: data,
         options: options
     });
+
     var legend = document.getElementById('legend');
+    var table = $('#trials').DataTable();
+    var selectedVal = 'Pending';
+    var date = new Date();
+    var startDate = new Date();
+    var endDate = new Date();
```
Comment (ChristianMurphy) [inline view/template/dashboard.hbs]: remove `date`, `startDate` and `endDate` from the global scope.

File: view/template/dashboard.hbs [hunk 2]
```diff
@@ -225,7 +229,63 @@
...
+            $('#selectedChoice').val('Completed');
+        }
+        table.draw();
+}
+);
+
+$.fn.dataTable.ext.search.push(
+      function (settings, data) {
+          'use strict';
+          startDate = data[4].split('/');
```
Comment (ChristianMurphy) [inline view/template/dashboard.hbs]: declare `date`, `startDate`, and `endDate` here instead of global.

File: view/template/dashboard.hbs [hunk 3]
```diff
@@ -225,7 +229,63 @@
...
+          startDate = new Date(parseInt(startDate[2], 10), parseInt(startDate[0], 10) - 1, parseInt(startDate[1], 10));
```
Comment (ChristianMurphy) [inline view/template/dashboard.hbs]: the date parsing done here for start date and end date could be replaced with.
``` js
var startDate = moment(data[4], 'MM-DD-YYYY');
var endDate = moment(data[5], 'MM-DD-YYYY');
```

File: view/template/dashboard.hbs [hunk 4]
```diff
@@ -225,7 +229,63 @@
...
+    var selectedVal = 'Pending';
+    var date = new Date();
+    var startDate = new Date();
+    var endDate = new Date();
 
     legend.innerHTML += radarChart.generateLegend();
+
+    $('.dropdown-menu').on('click', 'a', function () {
+        'use strict';
+        selectedVal = $(this).text();
```
Comment (ChristianMurphy) [inline view/template/dashboard.hbs]: `var selectedVal  = $(this).text().trim();` should work out any quirks in the string.
The set the value directly `$('#selectedChoice').val(selectedVal);`

File: view/template/dashboard.hbs [hunk 5]
```diff
@@ -225,7 +229,63 @@
...
+);
+
+$.fn.dataTable.ext.search.push(
+      function (settings, data) {
+          'use strict';
+          startDate = data[4].split('/');
+          startDate = new Date(parseInt(startDate[2], 10), parseInt(startDate[0], 10) - 1, parseInt(startDate[1], 10));
+          endDate = data[5].split('/');
+          endDate = new Date(parseInt(endDate[2], 10), parseInt(endDate[0], 10) - 1, parseInt(endDate[1], 10));
+          if ($('#selectedChoice').val().indexOf('All Trials') !== -1) {
```
Comment (ChristianMurphy) [inline view/template/dashboard.hbs]: replace `if` statements with a [switch statement](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch#Using_switch) testing against `$('#selectedChoice').val()`

**Comment by sylvr3** [comment]
@ChristianMurphy JSLint can't find "moment" in dashboard.hbs when I change the date parsing to the moment function call.

**Comment by sylvr3** [comment]
Also, I cannot declare any variables locally within a function because JSLint requires all vars to be declared at the top of the script.

**Comment by ChristianMurphy** [comment]
> JSLint can't find "moment" in dashboard.hbs when I change the date parsing to the moment function call.

After [line 11](https://github.com/ser515asu/PRP-Manhattan-Project/blob/master/view/.eslintrc.yml#L11) in `view/.eslintrc` add the line `moment: true`

> Also, I cannot declare any variables locally within a function because JSLint requires all vars to be declared at the top of the script.

ESLint checks that they are at the top of the global or a function scope.

``` javascript
$.fn.dataTable.ext.search.push(
    function (settings, data) {
        'use strict';
        var startDate = moment(data[4], 'MM-DD-YYYY');
        var endDate = moment(data[5], 'MM-DD-YYYY');
        ...
    }
);
```

will also pass the lint check.

**Comment by sylvr3** [comment]
I can't find eslintrc in the view folder

**Comment by ChristianMurphy** [comment]
On unix it will be a hidden file.
If you are trying to view it from file system, you will need to enable `show hidden files`

capture

From a terminal use `ls -A` to be able to see it.

**Comment by ChristianMurphy** [comment]
LGTM :ship:

**Pull request merged**